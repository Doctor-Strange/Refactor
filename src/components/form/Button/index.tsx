import React from "react";
import "./Button.scss";

const Button = (props: IButton) => {
  return (
    <button
    data-test-id="btn"
      className={props.class}
      disabled={props.disable || props.loading}
      onClick={() => props.click()}
    >
      {props.loading ? <span  data-test-id="Loading" className="Loading" /> : props.value}
    </button>
  );
};

interface IButton {
  value: string;
  class: string;
  loading: boolean;
  disable?: boolean;
  click: any;
}

export default Button;
