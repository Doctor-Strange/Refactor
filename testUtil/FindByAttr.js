

export const FindByAttr = (wrapper, value) => {
  return wrapper.find(`[data-test="${value}"]`);
};
