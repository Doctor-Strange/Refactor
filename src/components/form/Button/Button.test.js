import React from "react";
import { mount } from "enzyme";
import { FindByAttr } from "../../../../testUtil/FindByAttr";
import Button from "./";

const setup = (props) => {
  return mount(<Button {...props} />);
};

const clickMock = jest.fn();

const props = {
  value: "placeHolder",
  class: "BTN",
  loading: false,
  click: clickMock,
};

describe("Button component", () => {
  let wrapper;
  test("To Match Snapshot", () => {
    wrapper = setup(props);
    expect(wrapper).toMatchSnapshot();
  });

  describe("Props Attributes", () => {
    wrapper = setup(props);
    const button = FindByAttr(wrapper, "btn");
    const spinner = FindByAttr(wrapper, "Loading");
    test("has Class name", () => {
      expect(button.hasClass("BTN")).toEqual(true);
      expect(spinner.exists()).toEqual(false);
    });
  });

  describe("Show Spinner", () => {
    wrapper = setup({ ...props, loading: true });
    const spinner = FindByAttr(wrapper, "Loading");
    test("Show spinner if it's loading", () => {
      expect(spinner.exists()).toEqual(true);
    });
  });

  describe("Click mock simulator", () => {
    beforeEach(() => {
      clickMock.mockClear();
    });
    let button;
    test("disable click if it's loading", () => {
      wrapper = setup({ ...props, loading: true });
      button = FindByAttr(wrapper, "btn");
      button.simulate("click");
      expect(clickMock).toBeCalledTimes(0);
    });
    test("disable click if it's disable", () => {
      wrapper = setup({ ...props, disable: true });
      button = FindByAttr(wrapper, "btn");
      button.simulate("click");
      expect(clickMock).toBeCalledTimes(0);
    });
    test("disable click if it's loading", () => {
      wrapper = setup({ ...props, loading: false, disable: false });
      button = FindByAttr(wrapper, "btn");
      button.simulate("click");
      expect(button.text()).toEqual("placeHolder");
      expect(clickMock).toBeCalledTimes(1);
    });
  });
});
