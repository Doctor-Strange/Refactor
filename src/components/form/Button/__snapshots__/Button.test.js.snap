// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component To Match Snapshot 1`] = `ReactWrapper {}`;
